jina/optimize/__init__.py

```diff
@@ -0,0 +1,231 @@
+            parameters = yaml.load(file_path)
+            if "components" in parameters:
+                for i, component in enumerate(parameters["components"]):
+                    parameters["components"][i] = Optimizer._replace_param(
+                        component, trial_parameters
+                    )
+            parameters = Optimizer._replace_param(parameters, trial_parameters)
+            new_pod_file_path = trial_pod_dir / file_path.name
+            yaml.dump(parameters, open(new_pod_file_path, "w"))
+
+    def _create_trial_flows(self, trial_dir):
+        trial_flow_dir = trial_dir / "flows"
+        trial_flow_dir.mkdir(exist_ok=True)
+        yaml = YAML(typ="rt")
+        for file_path in [self.index_yaml, self.query_yaml]:
+            parameters = yaml.load(file_path)
+            for pod, val in parameters["pods"].items():
+                for pod_param in parameters["pods"][pod].keys():
+                    if pod_param.startswith("uses"):
+                        parameters["pods"][pod][pod_param] = str(
+                            trial_dir / self.pod_dir / Path(val[pod_param]).name
+                        )
+            trial_flow_file_path = trial_flow_dir / file_path.name
+            yaml.dump(parameters, open(trial_flow_file_path, "w"))
+
+    def _objective(self, trial):
+        trial_parameters = self._trial_parameter_sampler(trial)
+        trial_index_workspace = trial.workspace / "index"
+        trial_index_yaml = trial.workspace / "flows" / self.index_yaml.name
+        trial_query_yaml = trial.workspace / "flows" / self.query_yaml.name
+
+        if self.overwrite_trial_workspace:
+            self.flow_runner.clean_workdir(trial.workspace)
+
+        self._create_trial_pods(trial.workspace, trial_parameters)
+        self._create_trial_flows(trial.workspace)
+        self.flow_runner.run_indexing(trial_index_yaml, trial_index_workspace)
+        self.flow_runner.run_querying(trial_query_yaml, self.callback.process_result)
+
+        evaluation_values = self.callback.get_mean_evaluation()
+        op_name = list(evaluation_values)[0]
+        mean_eval = evaluation_values[op_name]
+        logger.info(colored(f"Avg {op_name}: {mean_eval}", "green"))
+        return mean_eval
+
+    def _export_params(self, study):
+        self.best_config_filepath.parent.mkdir(exist_ok=True)
+        yaml = YAML(typ="rt")
+        all_params = {**self.flow_runner.env_parameters, **study.best_trial.params}
+        yaml.dump(all_params, open(self.best_config_filepath, "w"))
+        logger.info(colored(f"Number of finished trials: {len(study.trials)}", "green"))
+        logger.info(colored(f"Best trial: {study.best_trial.params}", "green"))
+        logger.info(colored(f"Time to finish: {study.best_trial.duration}", "green"))
+
+    def optimize_flow(self, n_trials, direction="maximize", seed=42):
+        import optuna
+
+        sampler = optuna.samplers.TPESampler(seed=seed)
```

Comment:
what other samplers are available?

Reply:
There are many. I will keep an arg for this